Rakefile

```diff
@@ -1,9 +1,31 @@
 require 'rake'
 require 'rake/testtask'
 
-desc 'rake with no argument will run "rake test"'
-task default: :test
+require_relative 'lib/tasks/exercise'
+require_relative 'lib/tasks/exercise_test_tasks'
+require_relative 'lib/tasks/exercise_tests_runner'
+require_relative 'lib/tasks/rake_arguments'
 
-Rake::TestTask.new do |task|
-  task.pattern = 'test/**/*_test.rb'
+ExerciseTestTasks.new exercises: Exercise.all,
+                      options: RakeArguments.task_options
+
+task default: 'test'
+
+desc 'Run individual exercises or run all development and exercise tests'
+task :test do
+  next if RakeArguments.more_tasks?
+
+  Rake::Task['test:dev'].invoke
+  Rake::Task['test:exercises'].invoke
+end
+
+namespace :test do
+  desc 'Run all development tests located in the test directory'
+  Rake::TestTask.new :dev do |task|
+    task.options = RakeArguments.task_options
```

Comment:
This allows us to call `rake test:dev -- -p` and have it work. Previously it would only apply `-p` if you ran `TESTOPTS=-p rake test:dev`.

Also makes it use `-p` for `test:dev` tests when you run `rake test -- -p`